SignalServiceKit/src/Messages/OWSMessageReceiver.m

```diff
@@ -220,41 +222,43 @@ - (void)enqueueEnvelopeForProcessing:(OWSSignalServiceProtosEnvelope *)envelope
 
 - (void)drainQueue
 {
-    dispatch_async(self.class.serialGCDQueue, ^{
-        OWSMessageProcessingJob *_Nullable job = [self.finder nextJob];
-        if (job == nil) {
-            DDLogVerbose(@"%@ Queue is drained", self.tag);
-            return;
-        }
-
-        [self processJob:job
-              completion:^{
-                  [self drainQueue];
-              }];
-    });
+    AssertIsOnMainThread();
+
+    if (self.isDrainingQueue) {
+        return;
+    }
+    self.isDrainingQueue = YES;
+
+    [self drainQueueWorkStep];
 }
 
-- (void)processJob:(OWSMessageProcessingJob *)job completion:(void (^)())completion
+- (void)drainQueueWorkStep
 {
-    dispatch_async(dispatch_get_main_queue(), ^{
-        [self.messagesManager processEnvelope:job.envelopeProto
-                                   completion:^{
-                                       [self.finder removeJobWithId:job.uniqueId];
-                                       completion();
-                                   }];
-    });
-}
+    AssertIsOnMainThread();
+
+    OWSMessageProcessingJob *_Nullable job = [self.finder nextJob];
+    if (job == nil) {
+        self.isDrainingQueue = NO;
+        DDLogVerbose(@"%@ Queue is drained", self.tag);
+        return;
+    }
 
-#pragma mark Helpers
+    [self processJob:job
+          completion:^{
+              DDLogVerbose(@"%@ completed job. %lu jobs left.",
+                  self.tag,
+                  (unsigned long)[OWSMessageProcessingJob numberOfKeysInCollection]);
+              [self drainQueueWorkStep];
+          }];
+}
 
-+ (dispatch_queue_t)serialGCDQueue
+- (void)processJob:(OWSMessageProcessingJob *)job completion:(void (^)())completion
 {
-    static dispatch_once_t onceToken;
-    static dispatch_queue_t queue;
-    dispatch_once(&onceToken, ^{
-        queue = dispatch_queue_create("org.whispersystems.signal.messageProcessingQueue", NULL);
-    });
-    return queue;
+    [self.messagesManager processEnvelope:job.envelopeProto
```

Comment:
I'd love to move processing of envelopes off the main queue at some point. Thoughts?